src/utils/tokenCounter.ts

```diff
@@ -0,0 +1,125 @@
+import { get_encoding, Tiktoken } from "tiktoken"
+
+/**
+ * Shared token counter utility using tiktoken's cl100k_base encoding
+ * This provides a universal proxy token counter that works reasonably well for most models
+ */
+class TokenCounter {
+	private encoding: Tiktoken | null = null
+	private initializationPromise: Promise<void> | null = null
+
+	/**
+	 * Initialize the tiktoken encoding (lazy loading)
+	 */
+	private async initialize(): Promise<void> {
+		if (this.encoding) return
+
+		if (!this.initializationPromise) {
+			this.initializationPromise = this.doInitialize()
+		}
+		return this.initializationPromise
+	}
+
+	private async doInitialize(): Promise<void> {
+		try {
+			this.encoding = get_encoding("cl100k_base")
+		} catch (error) {
+			console.warn("Failed to initialize tiktoken encoding:", error)
+			this.encoding = null
+		}
+	}
+
+	/**
+	 * Estimate token count for a text string using tiktoken
+	 * Falls back to character-based estimation if tiktoken fails
+	 */
+	async estimateTokens(text: string): Promise<number> {
+		if (!text) return 0
+
+		try {
+			await this.initialize()
+			if (this.encoding) {
+				return this.encoding.encode(text).length
```

Comment:
Every time we are done with encoding we'll need to free the instance (this.encoder.free()) to release memory or it will could OOM iirc: I remember we ran into this issue that crashed the agent node process...